File: AutomationService/AutomationService/Features/Vorgaenge/Domain/Services/RegisterZeile.cs
using AutomationService.Features.RegisterHistorie.Domain.Services;

namespace AutomationService.Features.Vorgaenge.Domain.Services;

/// <summary>
/// Eine Zeile des Sachgebiete-Registers (§6.2) im verbindlichen Spaltenschema.
///
/// Vier Felder, nicht drei: Die gewachsene Kanzleidatei schiebt Zeichen,
/// Parteien und Sachbestand in <em>eine</em> Freitextzelle zusammen — genau die
/// Spalte, an der ein Parser des Altbestands scheitert. Der Export gibt sie
/// getrennt aus (#40); die App modelliert sie ohnehin getrennt.
/// </summary>
/// <param name="Jahr">
/// Vierstellig ("2026") — die Zwischenüberschrift, unter der die Zeile steht.
/// Abgeleitet und nie leer, damit keine Zeile aus der Gliederung fällt.
/// </param>
/// <param name="LaufendeNummer">
/// Die laufende Nummer im Jahrgang, null solange keine vergeben ist. Ein
/// Vorgang bekommt sie erst mit dem Abschluss.
/// </param>
/// <param name="Zeichen">Zeichen samt Abteilung ("01/26 C03").</param>
/// <param name="Parteien">"Mandant ./. Gegner".</param>
/// <param name="Sachbestand">"Sachverhalt v. 28.12.2025"; leer ohne Datum.</param>
/// <param name="Rechtsgebiet">Anzeigename des Sachgebiets ("Verkehrsrecht").</param>
/// <param name="Abgeschlossen">
/// Ob der Vorgang versendet und damit abgeschlossen ist. Steht in der Datei
/// nicht als eigene Spalte — dafür ist der Satzspiegel zu schmal —, sondern
/// zeichnet die Zeile aus (siehe RegisterDokument).
/// </param>
/// <param name="Quelle">
/// Ob die Zeile aus einem Vorgang der App oder aus der übernommenen
/// Registerhistorie stammt (<see cref="RegisterQuellen"/>). Vorbelegt mit
/// <c>vorgang</c>, damit die zweite Quelle keine einzige Aufrufstelle der
/// ersten anfassen musste.
/// </param>
/// <param name="HistorieId">Schlüssel der historischen Zeile — nur bei Quelle <c>historie</c>.</param>
/// <param name="VorgangReferenz">
/// Die Referenz des Vorgangs ("01/26 C03_HG-E 1427") — nur bei Quelle
/// <c>vorgang</c>. Damit kann die Ansicht die Zeile öffnen, ohne die Zeichen
/// wieder zerlegen zu müssen.
/// </param>
/// <param name="Sicherheit">
/// Wie sicher die Zerlegung der historischen Zeile war. Ein Vorgang der App ist
/// immer <c>hoch</c>: Seine Felder hat niemand aus einer Freitextzelle geraten.
/// </param>
/// <param name="Befunde">
/// Was an der historischen Zeile auffiel (Widerspruch, unbekanntes Kürzel).
/// Leer bei Vorgängen der App.
/// </param>
public sealed record RegisterZeile(
    string Jahr,
    int? LaufendeNummer,
    string Zeichen,
    string Parteien,
    string Sachbestand,
    string Rechtsgebiet,
    bool Abgeschlossen,
    string Quelle = RegisterQuellen.Vorgang,
    int? HistorieId = null,
    string? VorgangReferenz = null,
    string Sicherheit = RegisterSicherheiten.Hoch,
    IReadOnlyList<string>? Befunde = null);

File: Automation_App_Frontend/lib/features/vorgaenge/domain/entities/register_zeile.dart
import 'package:automation_app/features/vorgaenge/domain/entities/rechtsgebiet.dart';
import 'package:automation_app/features/vorgaenge/domain/entities/vorgang.dart';
import 'package:automation_app/features/vorgaenge/domain/entities/vorgang_status.dart';
import 'package:automation_app/features/vorgaenge/domain/services/vorgang_jahrgang.dart';
import 'package:equatable/equatable.dart';

/// Woher eine Registerzeile stammt. Muss zu `RegisterQuellen` im Backend
/// passen — die Werte stehen als Zeichenkette in der Antwort, nicht als Zahl.
abstract final class RegisterQuellen {
  static const String vorgang = 'vorgang';
  static const String historie = 'historie';
}

/// Wie sicher sich der Erzeuger der Importdatei bei einer historischen Zeile
/// war. Gegenstück zu `RegisterSicherheiten` im Backend.
///
/// Die Oberfläche sagt dieselben Wörter wie der Mandanten-Import
/// („sicher / unsicher / sehr unsicher") — zwei Wortleitern für dieselbe
/// Selbsteinschätzung wären eine zu viel.
abstract final class RegisterSicherheiten {
  static const String hoch = 'hoch';
  static const String mittel = 'mittel';
  static const String niedrig = 'niedrig';

  /// Wortlaut für den Bildschirm. Alles, was nicht ausdrücklich [hoch] sagt,
  /// gilt als prüfenswert — deshalb ist „ohne Angabe" der Rückfall.
  static String bezeichnung(String? wert) =>
      switch ((wert ?? '').trim().toLowerCase()) {
        RegisterSicherheiten.hoch => 'sicher',
        RegisterSicherheiten.mittel => 'unsicher',
        RegisterSicherheiten.niedrig => 'sehr unsicher',
        _ => 'ohne Angabe',
      };
}

/// Eine Zeile des Sachgebiete-Registers (§6.2), **gleich welcher Herkunft** —
/// ein laufender Vorgang der App oder eine übernommene Zeile aus dem
/// Registerbuch der Kanzlei.
///
/// Gebaut wird sie im Backend (`RegisterZeilenBau`), nicht hier. Das ist der
/// Kern von Issue #109: Vorher leitete die Ansicht ihre Zellen aus [Vorgang]
/// ab und die Word-/PDF-Datei ihre aus derselben Rechnung im Dienst — zwei
/// Quellen, die auseinanderlaufen konnten, ohne dass es jemandem auffiel.
/// Jetzt gibt es eine Quelle, und der Bildschirm zeigt, was in der Datei
/// steht. Nur [RegisterZeile.ausVorgang] rechnet noch selbst — dieselbe
/// Gefahr im Kleinen bewacht `vorgang_jahrgang_test.dart`.
class RegisterZeile extends Equatable {
  /// Vierstelliger Jahrgang („2019") — die Zwischenüberschrift, unter der die
  /// Zeile steht. Nie leer, damit keine Zeile aus der Gliederung fällt.
  final String jahr;

  /// Die laufende Nummer im Jahrgang; null, solange keine vergeben ist. Ein
  /// Vorgang bekommt sie erst mit dem Abschluss.
  final int? laufendeNummer;

  /// Zeichen samt Abteilung („01/26 C03") — nie die volle Referenz (§4.2).
  final String zeichen;

  /// Spalte „Sache": „Mandant ./. Gegner" (Form A) oder „Sachart Mandant"
  /// (Form B, wenn es keine Gegenseite gibt).
  final String parteien;

  /// „Unfall v. 28.12.2025"; leer, wenn nichts erfasst ist.
  final String sachbestand;

  final String rechtsgebiet;

  /// Ob der Vorgang abgeschlossen ist. Historische Zeilen sind es immer.
  final bool abgeschlossen;

  /// [RegisterQuellen.vorgang] oder [RegisterQuellen.historie].
  final String quelle;

  /// Schlüssel der historischen Zeile — nur bei Quelle „historie". Er ist die
  /// Adresse des Bearbeiten-Dialogs (`PUT /api/RegisterHistorie/{id}`).
  final int? historieId;

  /// Die Referenz des Vorgangs — nur bei Quelle „vorgang".
  final String? vorgangReferenz;

  final String sicherheit;

  /// Was an der historischen Zeile auffiel (Widerspruch, unbekanntes Kürzel).
  /// Leer bei Vorgängen der App.
  final List<String> befunde;

  const RegisterZeile({
    required this.jahr,
    required this.zeichen,
    this.laufendeNummer,
    this.parteien = '',
    this.sachbestand = '',
    this.rechtsgebiet = '',
    this.abgeschlossen = false,
    this.quelle = RegisterQuellen.vorgang,
    this.historieId,
    this.vorgangReferenz,
    this.sicherheit = RegisterSicherheiten.hoch,
    this.befunde = const [],
  });

  /// Ob die Zeile aus dem übernommenen Registerbuch stammt. Sie lässt sich
  /// nicht öffnen, sondern nur berichtigen — und trägt am Bildschirm immer den
  /// Status „Historie", damit sie von einem Vorgang der App unterscheidbar
  /// bleibt.
  bool get istHistorie => quelle == RegisterQuellen.historie;

  /// Ob die Zeile einen zweiten Blick verdient: Es gibt einen Befund, oder der
  /// Erzeuger der Datei war sich seiner Zerlegung nicht sicher.
  bool get zuPruefen =>
      befunde.isNotEmpty || sicherheit != RegisterSicherheiten.hoch;

  /// Die Selbsteinschätzung in Worten („sicher", „sehr unsicher").
  String get sicherheitText => RegisterSicherheiten.bezeichnung(sicherheit);

  /// Was in der vierten Spalte steht — ein nie erfasstes Sachgebiet wird zum
  /// Strich und nicht stillschweigend zu „Verkehrsrecht".
  String get rechtsgebietAnzeige => RechtsgebietWert.anzeige(rechtsgebiet);

  /// Beide Teile der Spalte „Sache" untereinander, wie auf schmalen Fenstern
  /// und im Word-Register.
  String get sacheUndSachbestand {
    if (sachbestand.isEmpty) return parteien;
    return parteien.isEmpty ? sachbestand : '$parteien\n$sachbestand';
  }

  factory RegisterZeile.fromJson(Map<String, dynamic> json) => RegisterZeile(
    jahr: json['jahr'] as String? ?? '',
    laufendeNummer: (json['laufendeNummer'] as num?)?.toInt(),
    zeichen: json['zeichen'] as String? ?? '',
    parteien: json['parteien'] as String? ?? '',
    sachbestand: json['sachbestand'] as String? ?? '',
    rechtsgebiet: json['rechtsgebiet'] as String? ?? '',
    abgeschlossen: json['abgeschlossen'] as bool? ?? false,
    quelle: json['quelle'] as String? ?? RegisterQuellen.vorgang,
    historieId: (json['historieId'] as num?)?.toInt(),
    vorgangReferenz: json['vorgangReferenz'] as String?,
    sicherheit: json['sicherheit'] as String? ?? RegisterSicherheiten.hoch,
    befunde: (json['befunde'] as List?)?.cast<String>() ?? const [],
  );

  /// Die Zeilen aus dem Umschlag `RegisterZeilenDto`. Ein Umschlag statt einer
  /// blanken Liste, damit später eine Gesamtzahl danebentreten kann, ohne dass
  /// der Vertrag seine Form ändert.
  static List<RegisterZeile> listeAusJson(Map<String, dynamic> json) => [
    for (final zeile in (json['zeilen'] as List? ?? const []))
      RegisterZeile.fromJson(zeile as Map<String, dynamic>),
  ];

  /// Baut eine Zeile aus einem bereits geladenen [Vorgang].
  ///
  /// **Nur für die Startseiten-Karte**, die den Vorgangsbestand ohnehin im
  /// Speicher hat und keinen zweiten Abruf rechtfertigt. Die Registeransicht
  /// selbst holt ihre Zeilen aus dem Backend — sie ist die Ansicht, die mit
  /// der Kanzleidatei übereinstimmen muss, und dort zählt eine Quelle.
  factory RegisterZeile.ausVorgang(Vorgang vorgang) => RegisterZeile(
    jahr: VorgangJahrgang.fuer(vorgang),
    laufendeNummer: vorgang.laufendeNummer,
    zeichen: vorgang.zeichen,
    parteien: vorgang.parteienBezeichnung,
    sachbestand: vorgang.registerSachbestand ?? '',
    rechtsgebiet: vorgang.rechtsgebiet,
    abgeschlossen: vorgang.status == VorgangStatus.versendet,
    vorgangReferenz: vorgang.referenz,
  );

  @override
  List<Object?> get props => [
    jahr,
    laufendeNummer,
    zeichen,
    parteien,
    sachbestand,
    rechtsgebiet,
    abgeschlossen,
    quelle,
    historieId,
    vorgangReferenz,
    sicherheit,
    befunde,
  ];
}
